docs/proposals/20240930-machine-drain-rules.md

```diff
@@ -0,0 +1,343 @@
+  * Allow to define if Pods should be excluded from drain
+* Use the current drain default rules as a fallback if neither `MachineDrainRule` CRs nor drain annotations
+  apply to a Pod (this also ensures compatibility with current behavior)
+* Stop waiting for volume detachment for Pods that are excluded from drain
+
+### Non-Goals
+
+* Change the drain behavior for DaemonSet and static Pods (we’ll continue to skip draining for both).
+  While the drain behavior itself won't be changed, we will stop waiting for detachment of volumes of DaemonSet Pods.
+
+### Future work
+
+* Align with the evolving landscape for Node drain in Kubernetes as soon as the new features are available, 
+  stable and can be used for the Kubernetes versions Cluster API supports (e.g. [KEP-4212](https://github.com/kubernetes/enhancements/issues/4212), [KEP-4563](https://github.com/kubernetes/enhancements/issues/4563)).
+
+  Because of the wide range of workload cluster Kubernetes versions supported in Cluster API (n-5), it will take 
+  significant time until we can use new Kubernetes features in all supported versions. So an implementation in
+  Cluster API is the only option to improve drain anytime soon (even for the oldest supported Kubernetes versions).
+
+  Also, please note that while we would like to align with Kubernetes and leverage its new features as soon as 
+  possible, there will always be a delta to be addressed in Cluster API due to the fact that Cluster API is operated 
+  from the management cluster, while drain is done in workload clusters.
+
+  Additionally, Cluster API should take care of many clusters, while Kubernetes features are by design scoped to a single cluster.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+#### `MachineDrainRule` CRD
+
+We propose to introduce a new `MachineDrainRule` namespace-scoped CRD. The goal of this CRD is to allow management 
+cluster admins/users to define a drain rule (a specific drain behavior) for a set of Pods.
+
+Additionally, drain rules can be restricted to a specific set of Machines, e.g. targeting only controlplane Machines, 
+Machines of specific Clusters, Linux Machines, etc.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDrainRule
+metadata:
+  name: example-rule
+  namespace: default
+spec:
+  drain:
+    # behavior can be Drain or Skip
+    behavior: Drain 
+    # order can only be set if behavior == Drain
+    # Pods with higher order are drained later
+    order: 100
+  machines:
+  - selector:
+      matchLabels:
+      matchExpressions:
+    clusterSelector:
+      matchLabels:
+      matchExpressions:
+  pods:
+  - selector:
+      matchLabels:
+      matchExpressions:
+    namespaceSelector:
+      matchLabels:
+      matchExpressions:
+```
+
+Notes:
+
+* All selectors are of type `metav1.LabelSelector`
+* In order to allow expressing selectors with AND and OR predicates, both `machines` and `pods` have a list of 
+  selectors instead of a single selector
```

Comment:
How does the AND/OR actually work? I guess entries in `matchExpression` is the `AND`, and then entries in the selectors list are OR'd? When the API is written up, this will want to be clear in the godoc

It may also be good to include an example of each AND'd and OR'd rules

Reply:
The plan is to use metav1.LabelSelector and corresponding utils. We were also planning to not change the semantic of it.

From the godoc of metav1.LabelSelector:
```go
// A label selector is a label query over a set of resources. The result of matchLabels and
// matchExpressions are ANDed. An empty label selector matches all objects. A null
// label selector matches no objects.
// +structType=atomic
type LabelSelector struct {
	// matchLabels is a map of {key,value} pairs. A single {key,value} in the matchLabels
	// map is equivalent to an element of matchExpressions, whose key field is "key", the
	// operator is "In", and the values array contains only "value". The requirements are ANDed.
	// +optional
	MatchLabels map[string]string `json:"matchLabels,omitempty" protobuf:"bytes,1,rep,name=matchLabels"`
	// matchExpressions is a list of label selector requirements. The requirements are ANDed.
	// +optional
	// +listType=atomic
	MatchExpressions []LabelSelectorRequirement `json:"matchExpressions,omitempty" protobuf:"bytes,2,rep,name=matchExpressions"`
}
```

So yes, within a single LabelSelector it's AND, and between multiple entires in the selector list it's OR

Reply:
Cool, sounds good, may just want to note that down somewhere, godoc in the API for the selectors slice for sure

Reply:
Yup, will update the proposal accordingly

Reply:
Also adjusted the example to explain this a bit better